includes/gateway/traits/sync-order-trait.php

```diff
@@ -0,0 +1,202 @@
+<?php
+
+trait Sync_Order
+{
+    /**
+     * Retrieve a charge by a given charge id (that attach to an order).
+     * Find some diff, then merge it back to WooCommerce system.
+     *
+     * @param  WC_Order $order WooCommerce's order object
+     *
+     * @return void
+     *
+     * @see    WC_Meta_Box_Order_Actions::save( $post_id, $post )
+     * @see    woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-actions.php
+     */
+    public function sync_payment($order)
+    {
+        $this->load_order($order);
+
+        try {
+            $charge = OmiseCharge::retrieve($this->get_charge_id_from_order());
+
+            /**
+             * Backward compatible with WooCommerce v2.x series
+             * This case is likely not going to happen anymore as this was provided back then
+             * when Omise-WooCommerce was introducing of adding charge.id into WC Order transaction id.
+             **/
+            if (!$this->order()->get_transaction_id()) {
+                $this->set_order_transaction_id($charge['id']);
+            }
+
+            switch ($charge['status']) {
+                case Omise_Payment::STATUS_SUCCESSFUL:
+                    $this->sync_order_handle_charge_successful($charge);
+                    break;
+                case Omise_Payment::STATUS_FAILED:
+                    $this->sync_order_handle_charge_failed($charge);
+                    break;
+                case Omise_Payment::STATUS_PENDING:
+                    $this->sync_order_handle_charge_pending();
+                    break;
+                case Omise_Payment::STATUS_EXPIRED:
+                    $this->sync_order_handle_charge_expired();
+                    break;
+                case Omise_Payment::STATUS_REVERSED:
+                    $this->sync_order_handle_charge_reversed();
+                    break;
+                default:
+                    throw new Exception(
+                        __('Cannot read the payment status. Please try sync again or contact Opn Payments support team at support@omise.co if you have any questions.', 'omise')
+                    );
+                    break;
+            }
+        } catch (Exception $e) {
+            $message = $this->allow_br('Opn Payments: Sync failed (manual sync).<br/>%s.');
+            $order->add_order_note(sprintf($message, $e->getMessage()));
+        }
+    }
+
+    /**
+     * This function handle successful charge, 
+     * and add order not when order is partially or full refunded 
+     * when sync order action was called
+     */
+    function sync_order_handle_charge_successful($charge)
```

Comment:
I see lots of methods don't have visibility. SonarCloud might throw more alerts for this. If it does. then let's add access modifiers.